### PR TITLE
Make the number of arguments of  `Hash#merge` variable

### PR DESCRIPTION
**Pull request by liwii**
# Abstract
Make the number of arguments of  `Hash#merge` variable.

# Background
In many websites such as  Stack Overflow and Qiita, many people are seeking how to merge more than three hashes. 

https://stackoverflow.com/questions/19548496/how-to-merge-multiple-hashes-in-ruby
https://qiita.com/hc0208/items/c662f5189fa383872f4e
https://stackoverrun.com/ja/q/4997431

Many ways, like using `Enumerable#inject` or calling `Hash#merge` multiple times, are proposed, but both don't seem intuitive. Especially when using block in `Hash#merge`, the code becomes too complicated.

# Proposal
Change the argument of `Hash#merge` from singular to variable length.
# Implementation
The code below.
# Evaluation
The code to merge more than three hashes became much simpler and more intuitive.

before
```ruby
hash1.merge(hash2).merge(hash3)

[hash1, hash2, hash3].inject do |result, part|
  result.merge(part) { |key, value1, value2| key + value1 + value2 }
end
```

after
```ruby
hash1.merge(hash2, hash3)
hash1.merge(hash2, hash3) { |key, value1, value2| key + value1 + value2 }
```
# Discussion
# Summary
The change is needed to make `Hash#merge` more useful and intuitive.

### CODE REVIEW

File: hash.c [hunk 1]
```diff
@@ -2572,17 +2572,19 @@ rb_hash_update_block_i(VALUE key, VALUE value, VALUE hash)
...
 static VALUE
-rb_hash_update(VALUE hash1, VALUE hash2)
+rb_hash_update(int argc, VALUE *argv, VALUE self)
 {
-    rb_hash_modify(hash1);
-    hash2 = to_hash(hash2);
-    if (rb_block_given_p()) {
-	rb_hash_foreach(hash2, rb_hash_update_block_i, hash1);
+    rb_hash_modify(self);
+    for(int i = 0; i < argc; i++){
```
Comment (knu) [inline hash.c]: CRuby is written in C90, so you should stick to the old syntax here.

```C
int i;

rb_hash_modify(self);
for (i = 0; i < argc; i++) {
```

File: hash.c [hunk 2]
```diff
@@ -2572,17 +2572,19 @@ rb_hash_update_block_i(VALUE key, VALUE value, VALUE hash)
...
+      VALUE hash = to_hash(argv[i]);
+      if (rb_block_given_p()) {
```
Comment (knu) [inline hash.c]: Maybe you can memoize this value.

**Comment by knu** [comment]
This proposal was accepted in yesterday's meeting, congratulations!

**Comment by liwii** [comment]
@knu 
Thanks for your review! I modified codes following the review, and added documents and test cases.

File: hash.c [hunk 3]
```diff
@@ -2641,28 +2678,43 @@ rb_hash_update_by(VALUE hash1, VALUE hash2, rb_hash_update_func *func)
...
  *  Returns a new hash containing the contents of <i>other_hash</i> and
  *  the contents of <i>hsh</i>. If no block is specified, the value for
  *  entries with duplicate keys will be that of <i>other_hash</i>. Otherwise
  *  the value for each duplicate key is determined by calling the block
  *  with the key, its value in <i>hsh</i> and its value in <i>other_hash</i>.
  *
+ *  More than one hash can be given to this method as arguments, then
+ *  the method returns a new hash containing the contents of the reciever
+ *  itself and all hashes given as arguments. The method also can be called 
+ *  with no argument, then the receiver itself will be returned;
```
Comment (nobu) [inline hash.c]: The result is a new hash, not the receiver **itself**.

**Comment by marcandre** [comment]
Nice. As far as documentation goes, I'd keep it simple. I don't think that discussing the trivial case of no argument is necessary. The interface should be kept to a minimum by basically changing `hsh.merge!(other_hash)` to `hsh.merge!(other_hash, another_hash, ...)`. See `Array#concat` for example.

**Comment by liwii** [comment]
@marcandre
Thanks for your review! I fixed documentation and simplified it. I'm not confident in English, so please tell me if you find something wrong in the documentation.

File: hash.c [hunk 4]
```diff
@@ -2548,41 +2548,62 @@ rb_hash_update_block_i(VALUE key, VALUE value, VALUE hash)
...
-    if (rb_block_given_p()) {
-	rb_hash_foreach(hash2, rb_hash_update_block_i, hash1);
-    }
-    else {
-	rb_hash_foreach(hash2, rb_hash_update_i, hash1);
+    int i;
+    bool block_given = rb_block_given_p();
+
+    rb_hash_modify(self);
+    for(i = 0; i < argc; i++){
```
Comment (nobu) [inline hash.c]: The indentation and style in block differ than others.

**Comment by liwii** [comment]
@nobu 
Thanks for review. I didn't notice it. I fixed the indentation in the new commit.

**Comment by liwii** [comment]
@simi 
Thanks for review. It's completely my mistake. I've already fixed it.

File: hash.c [hunk 5]
```diff
@@ -2548,41 +2548,62 @@ rb_hash_update_block_i(VALUE key, VALUE value, VALUE hash)
...
+ *  each _other_hash_. The method also can be called with no argument,
+ *  then nothing will change in the receiver.
+ *
+ *     h1 = { "a" => 100, "b" => 200 }
+ *     h1.merge!()     #=> {"a"=>100, "b"=>200}
+ *     h1              #=> {"a"=>100, "b"=>200}
  *
  *     h1 = { "a" => 100, "b" => 200 }
  *     h2 = { "b" => 254, "c" => 300 }
+ *     h3 = { "b" => 100, "d" => 400 }
```
Comment (mame) [inline hash.c]: Not needed?

**Comment by marcandre** [comment]
> @marcandre
> Thanks for your review! I fixed documentation and simplified it. I'm not confident in English, so please tell me if you find something wrong in the documentation.

Looks good. :+1: I'd even remove the part stating "The method also can be called with no argument, then nothing will change in the receiver.". It's important to test it, but it's not something people will use in general and should be obvious anyways.

File: hash.c [hunk 6]
```diff
@@ -2566,23 +2573,36 @@ rb_hash_update_block_i(VALUE key, VALUE value, VALUE hash)
...
+
+    rb_hash_modify(self);
+    for (i = 0; i < argc; i++){
+       VALUE hash = to_hash(argv[i]);
+       if (block_given) {
+          rb_hash_foreach(hash, rb_hash_update_block_i, self);
+       }
+       else {
+          rb_hash_foreach(hash, rb_hash_update_i, self);
+       }
```
Comment (k0kubun) [inline hash.c]: very trivial: These lines are indented with 3 spaces from `for` loop, but maybe usually it's indented with 4 spaces

**Comment by mame** [comment]
@liwii Merged!  Thank you, and I expect you to continue contribution to Ruby :smiley: 
@marcandre Oops, sorry I missed your comment.  I agree with you.  I've removed the sentence.  Thank you.

File: hash.c (hunk 6 above)
Comment (mame) [reply hash.c]: Fixed.  Thank you.

**Comment by marcandre** [comment]
> @marcandre Oops, sorry I missed your comment. I agree with you. I've removed the sentence. 

Yes, I commented a few minutes before your commit. Cool, thanks! I added a note in the `NEWS`.

Congrats @liwii  🎉

**Pull request closed without merging**